xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +929,691 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
+	backends := startTestServiceBackends(t, 2)
+
+	const clusterName = "cluster"
+	const locality1Weight = uint32(1)
+	const endpoint1Weight = uint32(1)
+	const locality2Weight = uint32(2)
+	const endpoint2Weight = uint32(2)
+	endpoints := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: clusterName,
+		Localities: []e2e.LocalityOptions{
+			{
+				Backends: []e2e.BackendOptions{{
+					Port:   testutils.ParsePort(t, backends[0]),
+					Weight: endpoint1Weight,
+				}},
+				Weight: locality1Weight,
+			},
+			{
+				Backends: []e2e.BackendOptions{{
+					Port:   testutils.ParsePort(t, backends[1]),
+					Weight: endpoint2Weight,
+				}},
+				Weight: locality2Weight,
+			},
+		},
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+	})
+	setRingHashLBPolicyWithHighMinRingSize(t, cluster)
+	route := e2e.DefaultRouteConfig("new_route", "test.server", clusterName)
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	const weight1 = endpoint1Weight * locality1Weight
+	const weight2 = endpoint2Weight * locality2Weight
+	const wantRPCs1 = float64(weight1) / float64(weight1+weight2)
+	const wantRPCs2 = float64(weight2) / float64(weight1+weight2)
+	numRPCs := computeIdealNumberOfRPCs(t, math.Min(wantRPCs1, wantRPCs2), errorTolerance)
+
+	// Send a large number of RPCs and check that they are distributed randomly.
+	gotPerBackend := checkRPCSendOK(ctx, t, client, numRPCs)
+	got := float64(gotPerBackend[backends[0]]) / float64(numRPCs)
+	if !cmp.Equal(got, wantRPCs1, cmpopts.EquateApprox(0, errorTolerance)) {
+		t.Errorf("Fraction of RPCs to backend %s: got %v, want %v (margin: +-%v)", backends[2], got, wantRPCs1, errorTolerance)
+	}
+	got = float64(gotPerBackend[backends[1]]) / float64(numRPCs)
+	if !cmp.Equal(got, wantRPCs2, cmpopts.EquateApprox(0, errorTolerance)) {
+		t.Errorf("Fraction of RPCs to backend %s: got %v, want %v (margin: +-%v)", backends[2], got, wantRPCs2, errorTolerance)
+	}
+}
+
+// Tests that ring hash policy that hashes using a fixed string ensures all RPCs
+// to go 1 particular backend; and that subsequent hashing policies are ignored
+// due to the setting of terminal.
+func (s) TestRingHash_FixedHashingTerminalPolicy(t *testing.T) {
```

Comment:
Doesn't this test cover whatever is covered in TestRingHash_UnsupportedHashPolicyUntilChannelIdHashing?

Reply:
That tests a different feature: While `TestRingHash_UnsupportedHashPolicyUntilChannelIdHashing` checks that unsupported policies do not apply, this one checks that a hash policy that come after a matching "terminal" policy does not apply (there was a confusing comments that stated something else, I removed it).

I think we should decide whether it's worth having the behaviors of the config selector that generates the hash be tested in those end to end tests or not. Happy to remove them if you think that the unit tests are sufficient. This one is also taken from c-core e2e tests.